vital/types/metadata.cxx

```diff
@@ -97,6 +97,22 @@ metadata
 ::metadata()
 { }
 
+// ---------------------------------------------------------------------
+metadata
+metadata
+::deep_copy()
```

Comment:
I wonder if this should instead be `clone( clone_type = clone_type::DEEP )`...

Reply:
I'd imagine so. I wasn't aware of this pattern.

Reply:
From what I can tell, a shallow copy is not really a thing here, since the `metadata` class takes ownership of its `metadata_item`s, so by necessity they would be copied as well. @mwoehlke-kitware Would you recommend still accepting the `clone_type` parameter and implementing the same behavior for both options, or just a rename from `deep_copy` to `clone`...?

Reply:
Would copying `m_metadata_map` not result in a shallow copy, since this is a `std::map< vital_metadata_tag, std::shared_ptr< metadata_item >`?

Reply:
That was my thought also; it seems like a shallow clone is just the copy ctor...

Reply:
Oh, yes, you're right. Oops.

Reply:
But consider `vital/types/metadata.h:371`. We don't handle both cases, and the case in which `item_ptr` is a `unique_ptr`, the copy constructor will fail.

Reply:
We might even consider deleting the copy constructor for the class altogether to prevent any code which would break when `item_ptr` is a `unique_ptr`

Reply:
Ah, right. Um... on second thought, maybe this should just *be* the copy ctor. Especially as the default ctor is broken (and being used in this PR!).

So, yes, please ensure the copy ctor is properly specified (i.e. either implemented or explicitly deleted). If you turn this into a copy ctor, please also explicitly default the move ctor and move assignment operator. (You might then delete the non-move assignment operator as not currently needed, perhaps with a note that it can be implemented if needed. Or you could implement it in terms of making a local copy and then move-assigning the copy to `this`. IOW, `T& op=(T const& o) { auto c = o; *this = std::move(c); return *this; }`.)

Reply:
Making this the copy ctor was my initial reaction. However, I was scared this could represent a breaking change that might be hard to track down across code unrelated to this PR. If you're comfortable with it, I'd agree that makes sense.

Reply:
(Oh, and there is precedent — within the metadata classes, even — for `clone` not taking a `clone_type`; if it doesn't, I believe that implies that only deep copies are possible. But it should return a pointer type. I'd prefer not a raw pointer, as that can lead to memory leaks, but there is unfortunately precedent for doing so.)

Reply:
Heh. On that note, there are a couple issues if the `unique_ptr` code was enabled. Which it *ought* to be, darn it; the use of `shared_ptr` is supposed to only be for broken STLs. I'm working on a PR for that. Meanwhile, yeah, I'm fine making this the copy ctor; any code currently using the copy ctor is currently broken. (Which includes *this PR*, but if you make this a copy ctor, that spot should naturally fix itself.)

Reply:
Forgot to close this; I believe everything should be addressed.